### ISSUE TITLE
ABCI: Avoid Amino encoding inside of Validator message

### ISSUE THREAD

**Issue by mdyring** [tendermint/tendermint#1524]
Per the below, it seems unnatural to pass Amino encoded data around inside Protobuf messages.

Until such a time that Amino sees wider adoption on major languages, even partially relying on Amino like this this will block developer adoption of Tendermint. 

I therefore submit as a *bug report* to:
Avoid Amino encoding inside the Validator message (types.proto). It seems this could be accomplished by extending the message to include the algorithm (Ed25519 etc). 

From Tendermint channel on Matrix:

```
ebuchman 18:30 
heads up on updated spec for Amino use in Tendermint (including examples for pubkeys and signatures): https://github.com/tendermint/tendermint/blob/develop/docs/specification/new-spec/encoding.md

mdyring 18:41
ebuchman: can you elaborate on how this will "spill over" to 3rd party applications not based on cosmos-sdk? that is, will this affect ABCI or IBC or is it simply an implementation detail in tendermint?

Zaki 18:41
We are going to have to support multiple encodings for IBC
Each "channel" will have to have a specifed encoding

mdyring 18:49
ok, yes I agree that IBC should support protobuf (as a first class citizen) as that is readily available to everybody.. great if it also supports amino
how about ABCI.. will that continue to be protobuf or also multiple formats now?

ebuchman 07:04
ABCI is still just protobuf but the pub_key byte array n the ResponseEndBlock is encoded in Amino for now
```

**Comment by ebuchman** [comment]
its really just a matter of appending 5 bytes to the pubkey. its the same 5 bytes every time: `1624DE6220`

**Comment by mdyring** [comment]
Hi Ethan,

I didn't realize it was as simple as that. But even if that continues to be the case, then ABCI (or any API for that matter) should strive to separate concerns.

From an ABCI consumer perspective, today I'd expect to provide 32 bytes of an Ed25519 pubkey in the Validator message. Having to prefix that with 5 magic bytes seems counter-intuitive and unexpected.

As an alternative to passing the pubkey algo inside the message, then as a quick(er) fix, perhaps these 5 bytes can be prefixed inside Tendermint?

If it is any comfort, I'd be making the same argument if XML, JSON or other formats started leaking into ABCI.

**Comment by ebuchman** [comment]
The problem is that we want to support other key types, which means we need to pass some information back in the protobuf message about the type, and we don't really want to support multiple typing schemes for the same object (eg. one in amino, one in protobuf)

**Comment by csviorama** [comment]
As long as it doesn’t end up like JWT :). Configurable algorithms can potentially introduce security issues. My question would be, how does it make the system more future proof as opposed to generally upgrading / replacing the algorithm when it’s required? For example if it’s necessary to be replaced due to security issues, then allowing usage of previous one would introduce issues.

**Comment by ebuchman** [comment]
What if we just did 

```
message PubKey {
  string type
  bytes data
}
```

And for now just support `type="ed25519"` and `data=<32-byte pubkey>`

**Comment by ebuchman** [comment]
That said, in the future, we'd probably do things like `type="amino"` and then `data=<amino encoded pubkey of potentially many types>`

**Comment by mdyring** [comment]
I am trying to understand what would the advantages be of doing it like this vs. the suggestion above?

AFAICS my suggestion:

* Addresses the issue of future extensibility and other key types, as it can express whatever you guys might decide to put into the "data" in Amino format later on.

* Provides a clean API and separation of concerns, which does not introduce new dependencies to ABCI developers.

* Requires a minimum of resource overhead (CPU and memory).

* Requires very few changes in Tendermint, which could be isolated inside a single function and would be easy to maintain.

IMHO moving opaque bytes around inside an API is a big red flag. It might make sense in some cases (like, passing file-level data around), but when the contained information is native to the API it should be expressed cleanly in the API and not encoded inside another format as a "Babushka API".

To me this an issue of code smell, which gives rise to a more overall concern of design decisions future and past. This concern is not limited to ABCI and is exacerbated by suggestions that IBC should only support Amino encoded proofs.

As longtime Tendermint/Cosmos fanboy (and heavily invested  by building a business for the ecosystem), I am extremely worried to see a move away from openness/interoperability. 

Introducing dependencies such as these add little value for application developers and delay projects. While it might make the Tendermint code slightly cleaner, first priority should always be interoperability.

**Comment by ebuchman** [comment]
The advantage is that we don't have to go all in on protobuf for our crypto types, which we're not planning to do. 

Protobuf is infectious on codebases and we'd like to minimize the extent to which it infiltrates.

Basically what I've proposed is the same as your single signer, and satisfies the basic requirement for the vast majority of users for now, so I would think it would be welcomed.

Using a `string` means we can be explicit about the type, rather than maintaining another lookup table.

Since the PubKey becomes a message type, we could always add more fields and effectively introduce your proposal at some point down the road, without having to commit to it now.

`oneof` makes a real mess of things, and was actually one of the primary motivators for Amino.

As for IBC, it will support multiple kinds of encoding schemes. When you open a connection with a chain, you will pick an encoding scheme, and that scheme lasts the duration of the connection. We're looking to support protobuf in there to make it as easy to use as possible.

Hope the above alleviates some of your concerns and you can get behind the PubKey proposal ! :)

**Comment by mdyring** [comment]
First of all, I am very happy to learn that IBC will support multiple encoding schemes. I am keeping my hopes up that Protobuf will be supported early on, so that we can launch our product without delay. :-)


With that out of the way, please excuse this Amino rant:

> Protobuf is infectious on codebases and we'd like to minimize the extent to which it infiltrates.

To an outsider, it seems Amino is equally infectious. Trouble is, that in addition to introducing yet another serialization format, there is no cross platform tools and the "bounties for implementations" program is not working. People looking to develop for Cosmos (outside of Go) are not likely to spend time on reinventing the wheel.

> `oneof` makes a real mess of things, and was actually one of the primary motivators for Amino.

I am not sure what the trouble is with `oneof` and Golang. From personal experience, I can report it works perfectly fine with C++, where one can conditionally check which type is stored. Java and JS as well, and supposedly it should work with Go these days: https://github.com/golang/protobuf/issues/29

The Amino idea of storing type information for each and every object large or small (say, a pubkey) to my mind makes very little sense and introduces overhead. In addition to some head-scratching: https://github.com/tendermint/go-crypto/issues/77 

When deserializing an object, the program of course knows what type(s) to expect, given its current state and point of execution. If a couple of options are available (like discussed above), this can trivially be expressed by wrapping these options inside `oneof`. Storing type information to supporting deserialization of any random type every time deserialization occurs is unneccesary (at best).


.. and now returning to the regular programming:

If threshold signatures are planned for the near future, I'd still argue for providing a proper interface in Protobuf. It is, simply put, professional courtesy to the 3rd party developers who are going to use ABCI.

> The advantage is that we don't have to go all in on protobuf for our crypto types, which we're not planning to do.

I think that is worth reconsidering, as the other other way to accomplish what you suggest is Amino-inside-Protobuf (a.k.a. Babushka API). This thinking will take the project down the path of Amino lock-in, adversely affecting adoption and running counter to the stated goals of the project.

With a development team of your size, working on much harder problems, the additional overhead should be negligible I believe.

I am painfully aware that the above contain a lot of negatives. So, just to be clear: I do appreciate the hard work you guys are doing and look very much forward to launch.

At the same time, I feel strongly about ensuring that Cosmos gets built in such a manner, that it will usable for a wide audience of developers. This to me is the largest part of the value proposition.

**Comment by amrali** [comment]
I've been following Amino's use in tendermint. The tradeoff of Amino being well integrated with tendermint code base while sacrificing wide adoption or dev support must not be made. If you truly believe that Amino is better than X or any other well supported FLOSS/OSS alternative out there, that case should be made independently from TM core, otherwise you unjustifiably end up fighting two uphill battles for developer adoption. If tendermint is widely successful, imagine the amount of support AIB has to deploy and maintain in terms of documentation, ports, community, etc. just to get people to use Amino.

Also imagine having TM use protobuffs or X that is widely supported and has many existing ports for many languages, even at the cost of sacrificing elegance (for a lack of a better word). Instead of asking questions, and spending days or weeks trying to figure out Amino, people would immediately focus on their application and business requirements which is precisely how TM success would be nourished and accelerated.

And this is not just causing pain points for IBC or ABCI, please have a look at tendermint/kms, Tony had to make ad-hoc parsing for Amino encoding in order to not delay development, he put explicit code comments where he decodes Amino messages to port Amino to Rust.

if you guys are absolutely certain that Amino is too good to not see the light, multiple encoding schemes is definitely the way to go, it's great to learn that IBC will support multiple encoding schemas, maybe we should have that for ABCI and any externally communicating part of TM as well (e.g., signers communicating to validators over a socket).

**Comment by ebuchman** [comment]
Thanks all. Please see more proposed changes to ABCI here: https://github.com/tendermint/abci/pull/245/files

We will also remove gogoprotobuf, unless there is any serious objection.

**Comment by ebuchman** [comment]
!! Check out https://github.com/tendermint/tendermint/issues/1625

**Comment by ebuchman** [comment]
This is basically done in https://github.com/tendermint/tendermint/pull/1667

Will be released in v0.20.0

Thanks everyone!!
